core/es_isolate.rs

```diff
@@ -0,0 +1,964 @@
+#[derive(Debug, Eq, PartialEq)]
+pub struct SourceCodeInfo {
+  pub code: String,
+  pub module_url_specified: String,
+  pub module_url_found: String,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum RecursiveLoadEvent {
+  Fetch(SourceCodeInfo),
+  Instantiate(ModuleId),
+}
+
+pub trait ImportStream: TryStream {
+  fn register(
+    &mut self,
+    source_code_info: SourceCodeInfo,
+    isolate: &mut EsIsolate,
+  ) -> Result<(), ErrBox>;
+}
+
+type DynImportStream = Box<
+  dyn ImportStream<
+      Ok = RecursiveLoadEvent,
+      Error = ErrBox,
+      Item = Result<RecursiveLoadEvent, ErrBox>,
+    > + Send
+    + Unpin,
+>;
+
+type DynImportFn = dyn Fn(DynImportId, &str, &str) -> DynImportStream;
+
+/// Wraps DynImportStream to include the DynImportId, so that it doesn't
+/// need to be exposed.
+#[derive(Debug)]
+struct DynImport {
+  pub id: DynImportId,
+  pub inner: DynImportStream,
+}
+
+impl fmt::Debug for DynImportStream {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    write!(f, "DynImportStream(..)")
+  }
+}
+
+impl Stream for DynImport {
+  type Item = Result<(DynImportId, RecursiveLoadEvent), (DynImportId, ErrBox)>;
+
+  fn poll_next(
+    self: Pin<&mut Self>,
+    cx: &mut Context,
+  ) -> Poll<Option<Self::Item>> {
+    let self_inner = self.get_mut();
+    match self_inner.inner.try_poll_next_unpin(cx) {
+      Poll::Ready(Some(Ok(event))) => {
+        Poll::Ready(Some(Ok((self_inner.id, event))))
+      }
+      Poll::Ready(None) => unreachable!(),
+      Poll::Ready(Some(Err(e))) => Poll::Ready(Some(Err((self_inner.id, e)))),
+      Poll::Pending => Poll::Pending,
+    }
+  }
+}
+
+impl ImportStream for DynImport {
+  fn register(
+    &mut self,
+    source_code_info: SourceCodeInfo,
+    isolate: &mut EsIsolate,
+  ) -> Result<(), ErrBox> {
+    self.inner.register(source_code_info, isolate)
+  }
+}
+
+pub struct ModuleInfo {
+  pub main: bool,
+  pub name: String,
+  pub handle: v8::Global<v8::Module>,
+  pub import_specifiers: Vec<String>,
+}
+
+/// A single execution context of JavaScript. Corresponds roughly to the "Web
+/// Worker" concept in the DOM. An Isolate is a Future that can be used with
+/// Tokio.  The Isolate future complete when there is an error or when all
+/// pending ops have completed.
+///
+/// Ops are created in JavaScript by calling Deno.core.dispatch(), and in Rust
+/// by implementing dispatcher function that takes control buffer and optional zero copy buffer
+/// as arguments. An async Op corresponds exactly to a Promise in JavaScript.
+#[allow(unused)]
+pub struct EsIsolate {
+  core_isolate: Box<Isolate>,
+
+  mods_: HashMap<ModuleId, ModuleInfo>,
+  pub(crate) next_dyn_import_id: DynImportId,
+  pub(crate) dyn_import_map:
+    HashMap<DynImportId, v8::Global<v8::PromiseResolver>>,
+  pub(crate) resolve_context: *mut c_void,
+
+  pending_dyn_imports: FuturesUnordered<StreamFuture<IntoStream<DynImport>>>,
+  dyn_import: Option<Arc<DynImportFn>>,
+  waker: AtomicWaker,
```

Comment:
Was AtomicWaker already part of deno_core::Isolate?

Reply:
Yes, it's used in two situations:
a) more ops to poll
b) more dyn imports to poll

Reply:
This doesn't need to be addressed now but feel like we should avoid tying deno_core to tokio. The AtomicWaker feels like it's unnecessarily growing the abstraction layer of deno_core.

Reply:
`AtomicWaker` is part of `futures` crate